File: HackerEarth/Holi And Colorful houses.cpp
#include<bits/stdc++.h>
#define maxx 1010
using namespace std;

string str;
int m,n;
int N,Q,t,len;

int main() {
    scanf("%d",&t);
    int cost1, cost2;

    while(t--) {
        cin>>N>>Q;
        cin>>str;
        len = str.length();

        for(int i=0; i < Q; i++) {
            cost1 = 0; cost2 = 0;
            cin>>m>>n;
            m -= 1;
            n -= 1;

            int st;
            /** traverse anti-clock-wise **/
            /** here we will get the cost1 due to traversing in anti-clock-wise **/
            st = m;
            while(st != n) {
                if(st == n) {
                    break;
                }
                if(st == 0) {
                    st = len-1;
                    if(str[st] != str[0]){
                        cost1++;
                    }
                } else {
                    st--;
                    if(str[st] != str[st+1]) {
                        cost1++;
                    }
                }
            }

            /** traverse clock-wise **/
            /** here we will get cost2 by traversing in clock-wise direction **/
            st = m;
            while(st != n) {
                if(st == n)
                    break;
                if(st == len-1) {
                    st=0;
                    if(str[0] != str[len-1]){
                        cost2++;
                    }
                } else {
                    st++;
                    if(str[st-1] != str[st]){
                        cost2++;
                    }
                }
            }

            cout<<min(cost1,cost2)<<endl;
        }
    }

    return 0;
}

File: HackerEarth/Holi_And_Colorful_houses.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Holi_And_Colorful_houses {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s = br.readLine();

        int t = Integer.parseInt(s);
        int N = 0, Q = 0;
        String p = "";
        int cost1 = 0, cost2 = 0;
        int st =0, m =0, n =0, len =0;

        while(t-- > 0) {
            String[] str = br.readLine().split(" ");
            N = Integer.parseInt(str[0]);
            Q = Integer.parseInt(str[1]);

            p = br.readLine();
            len = p.length();

            for(int i = 0; i < Q; i++) {
                cost1 = 0;
                cost2 = 0;
                str = br.readLine().split(" ");
                m = Integer.parseInt(str[0]);
                n = Integer.parseInt(str[1]);

                m -= 1;
                n -= 1;

                /** traverse anti-clock-wise **/
                /** here we will get the cost1 due to traversing in anti-clock-wise **/
                st = m;
                while(st != n) {
                    if(st == n)
                        break;
                    if(st == 0) {
                        st = len-1;
                        if(p.charAt(st) != p.charAt(0)) {
                            cost1++;
                        }
                    } else {
                        st--;
                        if(p.charAt(st) != p.charAt(st+1)){
                            cost1++;
                        }
                    }
                }

                st = m;
                while(st != n) {
                    if(st == n)
                        break;
                    if(st == len-1){
                        st = 0;
                        if(p.charAt(len-1) != p.charAt(0)){
                            cost2++;
                        }
                    } else {
                        st++;
                        if(p.charAt(st) != p.charAt(st-1)){
                            cost2++;
                        }
                    }
                }

                System.out.println(Math.min(cost1,cost2));
            }
        }
    }
}
